test/cpp/api/inference_mode.cpp

```diff
@@ -0,0 +1,286 @@
+#include <torch/script.h>
+#include <gtest/gtest.h>
+#include <test/cpp/api/support.h>
+
+using namespace torch::autograd;
+
+torch::Tensor functional_op(torch::Tensor& x) {
+  return x + x;
+}
+
+torch::Tensor& inplace_op(torch::Tensor& x) {
+  return x.add_(1);
+}
+
+torch::Tensor view_op(torch::Tensor& x) {
+  return x.view({2, 3});
+}
+
+/*
+  Only the following combos of Autograd & InplaceOrView keys on tensors are valid:
+    - Autograd=true, InplaceOrView=true
+    - Autograd=false, InplaceOrView=false
+*/
+void assert_tensor_dispatch_keys(torch::Tensor& x, bool Autograd, bool InplaceOrView) {
+  c10::DispatchKeySet ks = x.key_set();
+  ASSERT_EQ(ks.has(c10::DispatchKey::AutogradCPU), Autograd);
+  ASSERT_EQ(ks.has(c10::DispatchKey::InplaceOrView), InplaceOrView);
+}
+
+void assert_tensor_creation_meta(torch::Tensor& x, torch::autograd::CreationMeta creation_meta) {
+  ASSERT_EQ(static_cast<torch::autograd::DifferentiableViewMeta*>(x.unsafeGetTensorImpl()->autograd_meta())->get_creation_meta(), creation_meta);
+}
+
+void assert_TLS_states(bool inference_mode) {
+  ASSERT_EQ(InferenceMode::is_enabled(), inference_mode);
+  ASSERT_EQ(c10::impl::tls_is_dispatch_keyset_excluded(c10::autograd_dispatch_keyset), inference_mode);
+  ASSERT_EQ(c10::impl::tls_is_dispatch_key_included(c10::DispatchKey::InplaceOrView), !inference_mode);
+}
+
+TEST(InferenceModeTest, TestTLSState) {
+  assert_TLS_states(false);
+  {
+    InferenceMode guard(true);
+    assert_TLS_states(true);
+    {
+      InferenceMode guard(false);
+      assert_TLS_states(false);
+    }
+    assert_TLS_states(true);
+  }
+  assert_TLS_states(false);
+}
+
+TEST(InferenceModeTest, TestInferenceTensorCreation) {
+  {
+    InferenceMode guard(true);
+    // New tensor created in InferenceMode (inference tensor) doesn't have Autograd & InplaceOrView keys.
+    // Even if you ask it to requires_grad (which doesn't make sense in InferenceMode),
+    // it'll be silently ignored. This is not an error to make adding InferenceMode
+    // guard to existing code a smooth user experience.
+    torch::Tensor c = torch::ones({1, 2, 3});
+    assert_tensor_dispatch_keys(c, false, false);
+    ASSERT_FALSE(c.requires_grad());
+
+    torch::Tensor tmp = torch::ones({1, 2, 3}).set_requires_grad(true);
+    ASSERT_FALSE(tmp.requires_grad());
+    assert_tensor_dispatch_keys(tmp, false, false);
+
+    tmp = torch::ones({1, 2, 3}).set_requires_grad(false);
+    ASSERT_FALSE(tmp.requires_grad());
+    assert_tensor_dispatch_keys(tmp, false, false);
+  }
+}
+
+TEST(InferenceModeTest, TestTensorAssignment) {
+  torch::Tensor a = torch::ones({1, 2, 3}).set_requires_grad(true);
+  torch::Tensor b = torch::ones({1, 2, 3}).set_requires_grad(false);
+  torch::Tensor tmp, c;
+  // assignment operator preserves original keyset.
```

Comment:
This test seems kind of pointless because: "of course assignment operator preserves! It better, because Tensor is a shared pointer..." XD